File: frontend/curriculo_lattes_viewer/lib/src/controllers/trabalho_controller.dart
import 'dart:convert';

import 'package:curriculo_lattes_viewer/src/models/trabalho.dart';
import 'package:http/http.dart' as http;
import 'package:http/http.dart';

class TrabalhoController {
  final String _baseURL = 'http://localhost:5000/trabalho';

  Future<List<Trabalho>> listar() async {
    Response resposta = await http.get(Uri.parse('$_baseURL/listar'));
    var respostaJson = jsonDecode(resposta.body);
    List<Trabalho> trabalhos = [];
    for (var trabalhoJson in respostaJson) {
      trabalhos.add(Trabalho.fromJson(trabalhoJson));
    }
    return trabalhos;
  }

  Future<List<Trabalho>> listarRelacoes() async {
    Response resposta = await http.get(Uri.parse('$_baseURL/listar/relacoes'));
    var respostaJson = jsonDecode(resposta.body);
    List<Trabalho> trabalhos = [];
    for (var trabalhoJson in respostaJson) {
      trabalhos.add(Trabalho.fromJson(trabalhoJson));
    }
    return trabalhos;
  }

  Future<List<Trabalho>> filtrar(
      int? anoInicio,
      int? anoFim,
      int? idInstituto,
      String? idPesquisador,
      String? tipo,
      String? orderBy,
      String? sort,
      int posicaoInicial,
      int quantidadeItens) async {
    try {
      Response resposta = await http.get(Uri.parse(
          '$_baseURL/filtrar?anoInicio=$anoInicio&anoFim=$anoFim&idInstituto=$idInstituto&idPesquisador=$idPesquisador&tipo=$tipo&orderBy=$orderBy&sort=$sort&posicaoInicial=$posicaoInicial&quantidadeItens=$quantidadeItens'));
      var respostaJson = jsonDecode(resposta.body);
      List<Trabalho> trabalhos = [];
      for (var trabalhoJson in respostaJson) {
        trabalhos.add(Trabalho.fromJson(trabalhoJson));
      }
      return trabalhos;
    } catch (e) {
      print(e.toString());
      return [];
    }
  }

  Future<List<Trabalho>> filtrarAtualizado(
      int? anoInicio,
      int? anoFim,
      String? institutos,
      String? pesquisadores,
      String? tipo,
      String? orderBy,
      String? sort,
      int posicaoInicial,
      int quantidadeItens) async {
    try {
      Response resposta = await http.get(Uri.parse(
          '$_baseURL/filtrar/atualizado?anoInicio=$anoInicio&anoFim=$anoFim&institutos=$institutos&pesquisadores=$pesquisadores&tipo=$tipo&orderBy=$orderBy&sort=$sort&posicaoInicial=$posicaoInicial&quantidadeItens=$quantidadeItens'));
      var respostaJson = jsonDecode(resposta.body);
      List<Trabalho> trabalhos = [];
      for (var trabalhoJson in respostaJson) {
        trabalhos.add(Trabalho.fromJson(trabalhoJson));
      }
      return trabalhos;
    } catch (e) {
      print(e.toString());
      return [];
    }
  }

  Future<int> contar(
    int? anoInicio,
    int? anoFim,
    int? idInstituto,
    String? idPesquisador,
    String? tipo,
  ) async {
    try {
      var uri = Uri.parse(
          '$_baseURL/contar?anoInicio=$anoInicio&anoFim=$anoFim&idInstituto=$idInstituto&idPesquisador=$idPesquisador&tipo=$tipo');
      Response respostaRequisicao = await http.get(uri);
      var respostaJson = jsonDecode(respostaRequisicao.body);
      return respostaJson['totalTrabalhos'];
    } catch (e) {
      print(e.toString());
      return 0;
    }
  }

  Future<int> contarAtualizado(
    int? anoInicio,
    int? anoFim,
    String? institutos,
    String? pesquisadores,
    String? tipo,
  ) async {
    try {
      var uri = Uri.parse(
          '$_baseURL/contar/atualizado?anoInicio=$anoInicio&anoFim=$anoFim&institutos=$institutos&pesquisadores=$pesquisadores&tipo=$tipo');
      Response respostaRequisicao = await http.get(uri);
      var respostaJson = jsonDecode(respostaRequisicao.body);
      return respostaJson['totalTrabalhos'];
    } catch (e) {
      print(e.toString());
      return 0;
    }
  }
}

File: curriculo-lattes-backend/controller/trabalho_controller.py
import json
from flask import Blueprint
from flask import request
from flask_cors import cross_origin

from dao.trabalho_dao import TrabalhoDao

trabalho_blueprint = Blueprint('trabalho_blueprint', __name__)
trabalho_dao = TrabalhoDao()

@trabalho_blueprint.route('/listar', methods=['GET'])
@cross_origin()
def get_all():
    trabalhos = trabalho_dao.get_all()
    return json.dumps([ob.__dict__ for ob in trabalhos])

@trabalho_blueprint.route('/listar/relacoes', methods=['GET'])
@cross_origin()
def get_all_relacoes():
    trabalhos = trabalho_dao.get_all()
    resultado = []
    for trabalho in trabalhos:
        institutos = trabalho_dao.get_institutos_id(trabalho.id)
        pesquisadores = trabalho_dao.get_pesquisadores_id(trabalho.id)
        trabalho_dict = trabalho.__dict__
        trabalho_dict['institutos'] = institutos
        trabalho_dict['pesquisadores'] = pesquisadores
        resultado.append(trabalho_dict)

    return json.dumps(resultado)

@trabalho_blueprint.route('/filtrar', methods=['GET'])
@cross_origin()
def get_all_citacoes():
    try:
        ano_inicio = request.args.get('anoInicio')
        ano_fim = request.args.get('anoFim')
        id_instituto = request.args.get('idInstituto')
        id_pesquisador = request.args.get('idPesquisador')
        tipo = request.args.get('tipo')
        order_by = request.args.get('orderBy')
        sort = request.args.get('sort')
        posicao_inicial = request.args.get('posicaoInicial')
        quantidade_itens = request.args.get('quantidadeItens')
        trabalhos = trabalho_dao.filter(ano_inicio, ano_fim, id_instituto, id_pesquisador, tipo, order_by, sort, posicao_inicial, quantidade_itens)
        return trabalhos
    except Exception as e:
        print('exception')
        print(e)
        return json.dumps([])

@trabalho_blueprint.route('/filtrar/atualizado', methods=['GET'])
@cross_origin()
def get_all_citacoes_atualizado():
    try:
        ano_inicio = request.args.get('anoInicio')
        ano_fim = request.args.get('anoFim')
        institutos = request.args.get('institutos')
        pesquisadores = request.args.get('pesquisadores')
        tipo = request.args.get('tipo')
        order_by = request.args.get('orderBy')
        sort = request.args.get('sort')
        posicao_inicial = request.args.get('posicaoInicial')
        quantidade_itens = request.args.get('quantidadeItens')
        trabalhos = trabalho_dao.filter_atualizado(ano_inicio, ano_fim, institutos, pesquisadores, tipo, order_by, sort, posicao_inicial, quantidade_itens)
        return trabalhos
    except Exception as e:
        print('exception')
        print(e)
        return json.dumps([])

@trabalho_blueprint.route('/contar', methods=['GET'])
@cross_origin()
def count():
    try:
        ano_inicio = request.args.get('anoInicio')
        ano_fim = request.args.get('anoFim')
        id_instituto = request.args.get('idInstituto')
        id_pesquisador = request.args.get('idPesquisador')
        tipo = request.args.get('tipo')
        total_trabalhos = trabalho_dao.count(ano_inicio, ano_fim, id_instituto, id_pesquisador, tipo)
        return {'totalTrabalhos': total_trabalhos}
    except Exception as e:
        print('exception')
        print(e)
        return {'totalTrabalhos': 0}

@trabalho_blueprint.route('/contar/atualizado', methods=['GET'])
@cross_origin()
def count_atualizado():
    try:
        ano_inicio = request.args.get('anoInicio')
        ano_fim = request.args.get('anoFim')
        institutos = request.args.get('institutos')
        pesquisadores = request.args.get('pesquisadores')
        tipo = request.args.get('tipo')
        total_trabalhos = trabalho_dao.count_atualizado(ano_inicio, ano_fim, institutos, pesquisadores, tipo)
        return {'totalTrabalhos': total_trabalhos}
    except Exception as e:
        print('exception')
        print(e)
        return {'totalTrabalhos': 0}